### PR TITLE
[TPU Offload]Manage staging buffer in HBM

### PR DESCRIPTION
**Pull request by juncgu-google**
# Description

Add a staging buffer manager to book-keep the # of used staging blocks (page / chunk) for swap- in / out. This should be helpful in mitigating the OOM issue when kv cache offloading is handling very large of save / load blocks.

1. staging buffer size is controlled by env var: `TPU_OFFLOAD_STAGING_BUFFER_TOKENS=8192`. we will find a way to specify the buffer size in GB instead of in tokens.
2. when there are insufficient staging buffers for the incoming save / load operations, we will drop the extra save / load blocks.

# Tests

`pytest -sv tests/distrbuted/cpu_offloading_scheduler_test.py`

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/worker/tpu_worker_jax.py [hunk 1]
```diff
@@ -166,6 +166,28 @@ def determine_available_memory(self) -> int:
         total_hbm_limit_gb = round(total_hbm_limit / utils.GBYTES, 2)
         total_hbm_limit_cap_gb = round(total_hbm_limit_cap / utils.GBYTES, 2)
         total_hbm_used_gb = round(total_hbm_used / utils.GBYTES, 2)
+
+        if self.vllm_config.kv_transfer_config is not None:
+            kv_transfer_config = self.vllm_config.kv_transfer_config
+            if kv_transfer_config.kv_connector == "TPUConnector" and kv_transfer_config.kv_connector_module_path == "tpu_inference.distributed.tpu_connector_local":
+                # If kv offloading is enabled, we need to account for the memory used by the KV transfer buffer.
+                staging_buffer_tokens = int(
+                    os.getenv("TPU_OFFLOAD_STAGING_BUFFER_TOKENS", "8192"))
```
Comment (saikat-royc) [inline tpu_inference/worker/tpu_worker_jax.py]: For the end user, it may more more intuitive to configure bytes (rather than tokens) ?

we know the following:
1. runner.get_kv_cache_layout()
2. layers
3. block_size (ie num tokens).

File: tpu_inference/distributed/tpu_connector_local.py [hunk 2]
```diff
@@ -371,6 +441,152 @@ def get_finished(self,
...
+
+    def get_num_free_staging_blocks(self) -> int:
+        return self._num_free_blocks
+
+    def get_num_used_staging_blocks(self) -> int:
+        return self._num_blocks_for_load + self._num_blocks_for_save
+
+    def allocate(self, req_id: ReqId, num_blocks: int,
+                 usage: Literal["load", "save"]) -> int:
+        if num_blocks < 0:
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py:480]: this can be an assert

File: tpu_inference/distributed/tpu_connector_local.py [hunk 3]
```diff
@@ -371,6 +441,152 @@ def get_finished(self,
...
+        """
+        when num_finished_blocks is not given, we will assume the request is finished and should be removed.
+        """
+        num_freed_blocks = 0
+        # NOTE(jcgu): assuming FIFO execution order for a single request's save and
+        # load operations respectively
+        if usage == "load":
+            if req_id not in self._blocks_for_load:
+                # raise ValueError(f" there is no record of staging buffer (usage: {usage}) for Req:{req_id}")
+                logger.info(
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: logger.warning()

File: tpu_inference/distributed/tpu_connector_local.py [hunk 4]
```diff
@@ -463,6 +690,38 @@ def get_num_new_matched_tokens(
...
+            # still have something to load
+            if num_blocks_to_load > 0:
+                # NOTE(jcgu): fill real dst_blocks later when blocks get allocated.
+                dummy_dst_blocks = [-1] * num_blocks_to_load
+                self.load_specs[request.request_id] = LoadSpec(
+                    num_matched_tokens=num_matched_tokens,
+                    dst_blocks=dummy_dst_blocks,
+                    num_skip_leading_tokens=num_computed_tokens,
+                )
+                self.staging_buffer_manager.allocate(
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: let's put a note here there we assume the state of staging buffer pool to not change between get_num_free_staging_blocks (L697) and allocate() (L718), because there is one instance of the tpu connector scheduler expected to run

File: tpu_inference/distributed/tpu_connector_local.py [hunk 5]
```diff
@@ -371,6 +441,152 @@ def get_finished(self,
...
+    def get_num_used_staging_blocks(self) -> int:
+        return self._num_blocks_for_load + self._num_blocks_for_save
+
+    def allocate(self, req_id: ReqId, num_blocks: int,
+                 usage: Literal["load", "save"]) -> int:
+        if num_blocks < 0:
+            logger.warning(
+                f"  get {num_blocks} staging blocks to allocate for Req:{req_id}."
+            )
+            return
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: return -1

File: tpu_inference/distributed/tpu_connector_local.py (hunk 4 above)
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: we do not expect allocate to fail here. we can put an assert here for the return value to be equal to the num_blocks_to_load

File: tpu_inference/distributed/tpu_connector_local.py [hunk 6]
```diff
@@ -545,18 +795,30 @@ def update_state_after_alloc(self, request: "Request",
...
+
                 # adjust last partial block
-                last_partial_block_num_tokens = load_spec.num_matched_tokens - total_matched_blocks * self.block_size
-                need_last_block = self._adjust_last_partial_block(
-                    last_partial_block_num_tokens)
-                if need_last_block:
-                    total_matched_blocks += 1
-                assert total_matched_blocks <= len(
-                    all_blocks), f"{total_matched_blocks} > {len(all_blocks)}"
+                # total_matched_blocks = load_spec.num_matched_tokens // self.block_size
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: remove commented code

File: tpu_inference/distributed/tpu_connector_local.py [hunk 7]
```diff
@@ -656,22 +918,40 @@ def _prepare_req_meta(
...
+                    num_skip_leading_blocks:adjusted_num_total_blocks]
+                # This is a real save operation.
+                save_spec = SaveSpec(
+                    num_skip_leading_tokens=num_skip_leading_tokens,
+                    num_total_tokens=adjusted_num_total_tokens,
+                    is_final_save=is_finished,
+                    skip_save=False,
+                    src_blocks=src_block_ids,
+                )
+                self.staging_buffer_manager.allocate(
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: assert for succesful allocation = num_blocks_to_save

File: tpu_inference/distributed/tpu_connector_local.py [hunk 8]
```diff
@@ -656,22 +918,40 @@ def _prepare_req_meta(
...
+                    tracker.req_id,
+                    num_blocks=num_blocks_to_save,
+                    usage="save")
+                if adjusted_num_total_tokens > tracker.save_watermark:
+                    logger.info(
+                        f"      -> Old watermark {tracker.save_watermark}, new save_watermark count: {tracker.save_watermark}"
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: new save_watermark count {adjusted_num_total_tokens}

File: tpu_inference/distributed/tpu_connector_local.py [hunk 9]
```diff
@@ -683,6 +963,7 @@ def _prepare_req_meta(
                 is_final_save=True,
                 skip_save=True,
             )
+            # self.staging_buffer_manager.allocate(tracker.req_id, num_blocks=0, usage="save")
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: remove

**Comment by saikat-royc** [comment]
aligned with the high level appraoch on the block tracking, reporting and staging buffer budgeting. let's also add unit tests for added confidence in the PR

File: tpu_inference/worker/tpu_worker_jax.py (hunk 1 above)
Comment (juncgu-google) [reply tpu_inference/worker/tpu_worker_jax.py]: connector scheduler does not have the information of page_size_in_bytes directly. we will revise it later.

Comment (saikat-royc) [reply tpu_inference/worker/tpu_worker_jax.py]: ack

File: tpu_inference/distributed/tpu_connector_local.py [hunk 10]
```diff
@@ -451,6 +681,7 @@ def get_num_new_matched_tokens(
             )
             if self.cpu_backend.contains(key, pin_on_hit=True):
                 num_matched_tokens = end_token_idx
+                num_matched_blocks += 1
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py:686]: a comment here that each key represents a chunk which is equal to block size.

File: tpu_inference/distributed/tpu_connector_local.py [hunk 11]
```diff
@@ -463,6 +694,42 @@ def get_num_new_matched_tokens(
...
+            # still have something to load
+            if num_blocks_to_load > 0:
+                # NOTE(jcgu): fill real dst_blocks later when blocks get allocated.
+                dummy_dst_blocks = [-1] * num_blocks_to_load
+                self.load_specs[request.request_id] = LoadSpec(
+                    num_matched_tokens=num_matched_tokens,
+                    dst_blocks=dummy_dst_blocks,
+                    num_skip_leading_tokens=num_computed_tokens,
+                )
+                _num_allocated_blocks = self.staging_buffer_manager.allocate(
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py]: num_allocated_blocks (instead of _num_allocated_blocks)

File: tpu_inference/distributed/tpu_connector_local.py [hunk 12]
```diff
@@ -371,6 +442,152 @@ def get_finished(self,
...
+    """
+
+    def __init__(self, num_blocks: int):
+        self.num_blocks = num_blocks
+        # {req_id: list(num_occupied_staging_blocks)}
+        self._blocks_for_save: dict[ReqId, int] = {}
+        self._blocks_for_load: dict[ReqId, int] = {}
+
+        self._num_free_blocks: int = self.num_blocks
+        self._num_blocks_for_save: int = 0
```
Comment (saikat-royc) [inline tpu_inference/distributed/tpu_connector_local.py:469]: maybe put a comment that these two variables _num_blocks_for_load, _num_blocks_for_save track the actively used blocks across all requests for load and save respectively.

File: tests/distributed/cpu_offloading_scheduler_test.py [hunk 13]
```diff
@@ -0,0 +1,844 @@
...
+        offload_staging_buffer_tokens: int = -1,
+    ):
+        # update config
+        vllm_config = MockVllmConfig(block_size=block_size)
+        os.environ["TPU_OFFLOAD_DECODE_SAVE"] = str(offload_decode_save)
+        os.environ[
+            "TPU_OFFLOAD_PARTIAL_BLOCK_SAVE_BEHAVIOR"] = offload_partial_block_save_behavior
+        os.environ["TPU_OFFLOAD_PARTIAL_BLOCK_DYNAMIC_PAD_LOWER_LIMIT"] = str(
+            offload_partial_block_dynamic_pad_lower_limit)
+        logger.info(f"-------jcgu----------: {offload_staging_buffer_tokens}")
```
Comment (saikat-royc) [inline tests/distributed/cpu_offloading_scheduler_test.py]: remove -------jcgu----------

File: tests/distributed/cpu_offloading_scheduler_test.py [hunk 14]
```diff
@@ -0,0 +1,844 @@
...
+        num_matched_tokens = num_matched_blocks * scheduler.block_size
+        num_prompt_tokens = num_prompt_blocks * scheduler.block_size
+
+        prompt_tokens = list(range(num_prompt_tokens))
+        request = create_request("req1",
+                                 prompt_tokens,
+                                 block_size=scheduler.block_size)
+
+        # Simulate a cache hit for the first 3 block
+        keys_gen = scheduler.token_processor.process_tokens(prompt_tokens)
```
Comment (saikat-royc) [inline tests/distributed/cpu_offloading_scheduler_test.py]: this block L272-276 can be a helper function used by multiple test_* functions

File: tpu_inference/distributed/tpu_connector_local.py (hunk 2 above)
Comment (juncgu-google) [reply tpu_inference/distributed/tpu_connector_local.py:480]: put a warning and return the invalid num_blocks. let the connector scheduler to make the decision (assert num_allocated_blocks >=0).

**Comment by juncgu-google** [comment]
Thanks for the reviews!
